wire-ios-sync-engine/Source/Use cases/CreateTeamOneOnOneConversationUseCase.swift

```diff
@@ -0,0 +1,154 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+protocol CreateTeamOneOnOneConversationUseCaseProtocol {
+
+    func invoke(
+        with user: ZMUser,
+        syncContext: NSManagedObjectContext
+    ) async throws -> NSManagedObjectID
+
+}
+
+public enum CreateTeamOneOnOneConversationError: Error {
+
+    case userDoesNotExist
+    case userIsNotOnSameTeam
+    case missingUserQualifiedID
+    case mlsMigratorNotFound
+    case failedToCreateMLSConversation(Error)
+    case failedToCreateProteusConversation(Error)
+    case noCommonProtocols
+    case conversationNotFound
+
+}
+
+/// Creates the team one on one conversation with a particlar user, depending
+/// on the currently supported protocols of the self and other users.
+
+struct CreateTeamOneOnOneConversationUseCase: CreateTeamOneOnOneConversationUseCaseProtocol {
+
+    private typealias Error = CreateTeamOneOnOneConversationError
+
+    private let protocolSelector: OneOnOneProtocolSelectorInterface
+    private let migrator: OneOnOneMigratorInterface?
+    private let service: ConversationServiceInterface
+
+    init(
+        protocolSelector: OneOnOneProtocolSelectorInterface,
+        migrator: OneOnOneMigratorInterface?,
+        service: ConversationServiceInterface
+    ) {
+        self.protocolSelector = protocolSelector
+        self.migrator = migrator
+        self.service = service
+    }
+
+    func invoke(
+        with user: ZMUser,
+        syncContext: NSManagedObjectContext
+    ) async throws -> NSManagedObjectID {
+        let userID = try await syncContext.perform {
+            guard user.isOnSameTeam(otherUser: ZMUser.selfUser(in: syncContext)) else {
+                throw Error.userIsNotOnSameTeam
+            }
+
+            guard
+                let userID = user.remoteIdentifier,
+                let domain = user.domain ?? BackendInfo.domain
+            else {
+                throw Error.missingUserQualifiedID
+            }
+
+            return QualifiedID(
+                uuid: userID,
+                domain: domain
+            )
+        }
+
+        switch await protocolSelector.getProtocolForUser(
+            with: userID,
+            in: syncContext
+        ) {
+        case .mls:
+            return try await createMLSConversation(
+                userID: userID,
+                in: syncContext
+            )
+
+        case .proteus:
+            return try await createProteusConversation(
+                with: user,
+                in: syncContext
+            )
+
+        case .none, .mixed:
+            throw Error.noCommonProtocols
```

Comment:
If there are no common protocols it will return nil. The "mixed" protocol doesn't really make sense here, I think it happens now because previously there were two distinct message protocol types and they got merged into one. One of them was for the conversation protocol, the other was for which protocols a user supports.

Reply:
ok thks for the explanation